libbpf-tools/hardirqs.bpf.c

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "hardirqs.h"
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+#define MAX_ENTRIES	256
+
+const volatile bool targ_dist = false;
+const volatile bool targ_ns = false;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct irq_key);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} irqs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, u32);
+	__type(value, u64);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct irq_key);
+	__type(value, struct hist);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} hists SEC(".maps");
+
+static struct hist zero_hist;
+
+SEC("tracepoint/irq/irq_handler_entry")
+int handle__irq_handler(struct trace_event_raw_irq_handler_entry *ctx)
+{
+	struct irq_key key = {};
+	u64 *count, zero = 0;
+
+	bpf_probe_read_kernel_str(&key.name, sizeof(key.name), ctx->__data);
+	count = bpf_map_lookup_or_try_init(&irqs, &key, &zero);
+	if (!count)
+		return 0;
+	*count += 1;
```

Comment:
should be atomic?

Reply:
In my mind, only one instance of the irq handler would be running on one of the processes.  So I think we don't need atomic?

Reply:
I'm not an expert on how IRQs work and how they are handled in Linux, so don't know. But my assumption was that each CPU can handle them independently. But again, no idea.